avocado/core/mux.py

```diff
@@ -29,6 +36,156 @@
 REMOVE_VALUE = 1
 
 
+class MuxTree(object):
+
+    """
+    Object representing part of the tree from the root to leaves or another
+    multiplex domain. Recursively it creates multiplexed variants of the full
+    tree.
+    """
+
+    def __init__(self, root):
+        """
+        :param root: Root of this tree slice
+        """
+        self.pools = []
+        for node in self._iter_mux_leaves(root):
+            if node.is_leaf:
+                self.pools.append(node)
+            else:
+                self.pools.append([MuxTree(child) for child in node.children])
+
+    @staticmethod
+    def _iter_mux_leaves(node):
+        """ yield leaves or muxes of the tree """
+        queue = collections.deque()
+        while node is not None:
+            if node.is_leaf or getattr(node, "multiplex", None):
+                yield node
+            else:
+                queue.extendleft(reversed(node.children))
+            try:
+                node = queue.popleft()
+            except IndexError:
+                raise StopIteration
+
+    def __iter__(self):
+        """
+        Iterates through variants
+        """
+        pools = []
+        for pool in self.pools:
+            if isinstance(pool, list):
+                pools.append(itertools.chain(*pool))
+            else:
+                pools.append(pool)
+        pools = itertools.product(*pools)
+        while True:
+            # TODO: Implement 2nd level filters here
+            # TODO: This part takes most of the time, optimize it
+            yield list(itertools.chain(*pools.next()))
+
+
+class MuxPlugin(object):
+    """
+    Base implementation of Mux-like Varianter plugin. It should be used as
+    a base class in conjunction with
+    :class:`avocado.core.plugin_interfaces.VarianterPlugin`.
+    """
+    root = None
+    variants = None
+    default_params = None
+    mux_path = None
+    debug = None
+
+    def initialize_mux(self, root, mux_path, debug):
+        """
+        Initialize the basic values
+
+        :note: We can't use __init__ as this object is intended to be used
```

Comment:
I'm not sure what are you trying to suggest. If I understand the `invoke_kwds` correctly it allows to pass the kwargs to the `__init__`, but during the dispatcher's `__init__` I don't have the `args` available yet. This happens before `Varianter.parse`.

Could you please explain a bit deeper what you had in mind?